dnsmasq_interface.c

```diff
@@ -396,96 +393,72 @@ void FTL_reply(unsigned short flags, char *name, struct all_addr *addr, int id)
 	struct timeval response;
 	gettimeofday(&response, 0);
 
-	if(flags & F_CONFIG)
-	{
-		// Answered from local configuration, might be a wildcard or user-provided
-		// Save status in corresponding query identified by dnsmasq's ID
-		bool found = false;
-		int i;
-
-		// Search match in known queries
-		// See comments in FTL_forwarded() for further details about this loop
-		validate_access("queries", counters.queries-1, false, __LINE__, __FUNCTION__, __FILE__);
-		int until = MAX(0, counters.queries-MAXITER);
-		for(i = counters.queries-1; i >= until; i--)
-		{
-			// Check UUID of this query
-			if(queries[i].id == id)
-			{
-				found = true;
-				break;
-			}
-		}
+	// Save status in corresponding query identified by dnsmasq's ID
+	bool found = false;
+	int i;
 
-		if(!found)
+	// Search match in known queries
+	// See comments in FTL_forwarded() for further details about this loop
+	validate_access("queries", counters.queries-1, false, __LINE__, __FUNCTION__, __FILE__);
+	int until = MAX(0, counters.queries-MAXITER);
+	for(i = counters.queries-1; i >= until; i--)
+	{
+		// Check UUID of this query
+		if(queries[i].id == id)
 		{
-			// This may happen e.g. if the original query was a PTR query or "pi.hole"
-			// as we ignore them altogether
-			disable_thread_lock();
-			return;
+			found = true;
+			break;
 		}
+	}
 
-		if(!queries[i].complete)
-		{
-			// This query is no longer unknown
-			counters.unknown--;
-			// Answered from a custom (user provided) cache file
-			counters.cached++;
+	if(!found)
+	{
+		// This may happen e.g. if the original query was a PTR query or "pi.hole"
+		// as we ignore them altogether
+		if(debug) logg("FTL_reply(): Query %i has not been found", id);
+		disable_thread_lock();
+		return;
+	}
 
-			if(strcmp(answer, "(NXDOMAIN)") == 0 ||
-			   strcmp(answer, "0.0.0.0") == 0 ||
-			   strcmp(answer, "::") == 0)
-				queries[i].status = QUERY_WILDCARD;
-			else
-				queries[i].status = QUERY_CACHE;
+	if(queries[i].reply != REPLY_UNKNOWN)
```

Comment:
Would checking `queries[i].complete` do the same thing? (that is what is used in the old code)

Reply:
No, not in the following case:

1. FTL receives a query (FTL creates `queries[i]`)
2. `dnsmasq` decides to forward it (FTL sets `queries[i].status = QUERY_FORWARDED` and queries[i].complete = true` as the query status is now known)
3. When we get the reply from upstream we have to check if the `reply` is already known (this is this code)